File: WinForms/CS/LayoutMainDemo/Modules/XtraLayoutPanel/Table/DetailView.cs
using System;
using System.ComponentModel;
using System.Drawing;
using System.Windows.Forms;
using DevExpress.Utils;
using DevExpress.Utils.Layout;
using DevExpress.Utils.Drawing;
using DevExpress.Utils.Extensions;
using DevExpress.DXperience.Demos;

namespace DevExpress.XtraLayout.Demos.Modules.XtraLayoutPanel {
    public partial class DetailView : XtraLayoutPanelDemoViewBase {
        Rectangle _newColumnButtonRect;
        Rectangle _newRowButtonRect;
        ObjectState _newRowButtonState;
        ObjectState _newColumnButtonState;

        public DetailView() {
            InitializeComponent();
        }
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
        public bool ShowGrid {
            get { return tablePanel.ShowGrid == DefaultBoolean.True; }
            set {
                tablePanel.ShowGrid = value ? DefaultBoolean.True : DefaultBoolean.False;
                InvalidatePanel();
            }
        }
        static readonly object markerObj = new object();
        protected override void InitializeControls() {
            base.InitializeControls();
            deHireDate.DateTime = TutorialConstants.Now - TimeSpan.FromDays(365 * 5);
            deBirthDate.DateTime = TutorialConstants.Now - TimeSpan.FromDays(365 * 35);
            tablePanel.ForEachTableEntity(x => x.Tag = markerObj);
        }
        public override void ResetLayout() {
            LayoutPanel.SuspendLayout();
            try {
                base.ResetLayout();
                for(int n = tablePanel.Rows.Count - 1; n >= 0; n--) {
                    if(!ReferenceEquals(tablePanel.Rows[n].Tag, markerObj)) tablePanel.Rows.RemoveAt(n);
                }
                for(int n = tablePanel.Columns.Count - 1; n >= 0; n--) {
                    if(!ReferenceEquals(tablePanel.Columns[n].Tag, markerObj)) tablePanel.Columns.RemoveAt(n);
                }
            }
            finally {
                LayoutPanel.ResumeLayout();
            }

        }

        #region Rects

        Rectangle NewColumnButtonRect {
            get { return _newColumnButtonRect; }
            set {
                if(NewColumnButtonRect == value) return;
                _newColumnButtonRect = value;
                InvalidatePanel();
            }
        }
        Rectangle NewRowButtonRect {
            get { return _newRowButtonRect; }
            set {
                if(NewRowButtonRect == value) return;
                _newRowButtonRect = value;
                InvalidatePanel();
            }
        }
        ObjectState NewColumnButtonState {
            get { return _newColumnButtonState; }
            set {
                if(NewColumnButtonState == value) return;
                _newColumnButtonState = value;
                InvalidatePanel();
            }
        }
        ObjectState NewRowButtonState {
            get { return _newRowButtonState; }
            set {
                if(NewRowButtonState == value) return;
                _newRowButtonState = value;
                InvalidatePanel();
            }
        }

        #endregion

        #region Paint

        void OnTablePanelPaint(object sender, PaintEventArgs e) {
            CalculateRectangles();
            using(GraphicsCache cache = new GraphicsCache(e)) {
                if(ShowGrid) DrawButtons(cache);
            }
        }
        private void DrawButtons(GraphicsCache cache) {
            cache.DrawImageUnscaled(CreateColumnImage, NewColumnButtonRect);
            cache.DrawImageUnscaled(CreateRowImage, NewRowButtonRect);
        }

        #endregion

        #region Events

        void OnTablePanelMouseUp(object sender, MouseEventArgs e) {
            CalculateObjectStates(e);
        }
        void OnTablePanelMouseDown(object sender, MouseEventArgs e) {
            CalculateObjectStates(e);
        }
        void OnTablePanelMouseMove(object sender, MouseEventArgs e) {
            CalculateRectangles();
            CalculateObjectStates(e);
        }
        void OnTablePanelMouseClick(object sender, MouseEventArgs e) {
            if(e.Button != MouseButtons.Left) return;

            if(NewColumnButtonRect.Contains(e.Location)) {
                CreateColumn();
            }
            else if(NewRowButtonRect.Contains(e.Location)) {
                CreateRow();
            }
        }

        #endregion

        #region Calculations

        static readonly Size ButtonSize = ScaleUtils.ScaleValue(new Size(16, 16));

        private void CalculateRectangles() {
            NewColumnButtonRect = LayoutPanel.Bounds.RightTop().CreateRect(ButtonSize).WithOffset(-ButtonSize.Width, ButtonSize.Height);
            NewRowButtonRect = LayoutPanel.Bounds.RightBottom().CreateRect(ButtonSize).WithOffset(-ButtonSize.Width, -ButtonSize.Height);
        }
        private void CalculateObjectStates(MouseEventArgs e) {
            NewColumnButtonState = CalculateObjectStateCore(NewColumnButtonRect, e);
            NewRowButtonState = CalculateObjectStateCore(NewRowButtonRect, e);
        }
        private ObjectState CalculateObjectStateCore(Rectangle bounds, MouseEventArgs e) {
            return bounds.Contains(e.Location) && (e.Button & MouseButtons.Left) == 0 ? ObjectState.Hot : ObjectState.Normal;
        }

        #endregion

        private void CreateColumn() {
            tablePanel.Columns.Add(new TablePanelColumn(TablePanelEntityStyle.Relative, 50f));
        }
        private void CreateRow() {
            tablePanel.Rows.Add(new TablePanelRow(TablePanelEntityStyle.Absolute, 38f));
        }
        private void InvalidatePanel() {
            LayoutPanel.Invalidate();
        }
        private Image CreateColumnImage {
            get { return imageList.GetImage(NewColumnButtonState == ObjectState.Hot ? 1 : 0, new Size(16, 16)); }
        }
        private Image CreateRowImage {
            get { return imageList.GetImage(NewRowButtonState == ObjectState.Hot ? 3 : 2, new Size(16, 16)); }
        }
        public override XtraLayoutPanelBase LayoutPanel { get { return tablePanel; } }
        public TablePanel TableLayoutPanel { get { return tablePanel; } }
    }
}

File: WinForms/VB/LayoutMainDemo/Modules/XtraLayoutPanel/Table/DetailView.vb
Imports System
Imports System.ComponentModel
Imports System.Drawing
Imports System.Windows.Forms
Imports DevExpress.Utils
Imports DevExpress.Utils.Layout
Imports DevExpress.Utils.Drawing
Imports DevExpress.Utils.Extensions
Imports DevExpress.DXperience.Demos

Namespace DevExpress.XtraLayout.Demos.Modules.XtraLayoutPanel

    Public Partial Class DetailView
        Inherits XtraLayoutPanelDemoViewBase

        Private _newColumnButtonRect As Rectangle

        Private _newRowButtonRect As Rectangle

        Private _newRowButtonState As ObjectState

        Private _newColumnButtonState As ObjectState

        Public Sub New()
            InitializeComponent()
        End Sub

        <DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)>
        Public Property ShowGrid As Boolean
            Get
                Return tablePanel.ShowGrid = DefaultBoolean.True
            End Get

            Set(ByVal value As Boolean)
                tablePanel.ShowGrid = If(value, DefaultBoolean.True, DefaultBoolean.False)
                InvalidatePanel()
            End Set
        End Property

        Private Shared ReadOnly markerObj As Object = New Object()

        Protected Overrides Sub InitializeControls()
            MyBase.InitializeControls()
            deHireDate.DateTime = TutorialConstants.Now - TimeSpan.FromDays(365 * 5)
            deBirthDate.DateTime = TutorialConstants.Now - TimeSpan.FromDays(365 * 35)
            tablePanel.ForEachTableEntity(Sub(x) x.Tag = markerObj)
        End Sub

        Public Overrides Sub ResetLayout()
            LayoutPanel.SuspendLayout()
            Try
                MyBase.ResetLayout()
                For n As Integer = tablePanel.Rows.Count - 1 To 0 Step -1
                    If Not ReferenceEquals(tablePanel.Rows(n).Tag, markerObj) Then tablePanel.Rows.RemoveAt(n)
                Next

                For n As Integer = tablePanel.Columns.Count - 1 To 0 Step -1
                    If Not ReferenceEquals(tablePanel.Columns(n).Tag, markerObj) Then tablePanel.Columns.RemoveAt(n)
                Next
            Finally
                LayoutPanel.ResumeLayout()
            End Try
        End Sub

#Region "Rects"
        Private Property NewColumnButtonRect As Rectangle
            Get
                Return _newColumnButtonRect
            End Get

            Set(ByVal value As Rectangle)
                If NewColumnButtonRect = value Then Return
                _newColumnButtonRect = value
                InvalidatePanel()
            End Set
        End Property

        Private Property NewRowButtonRect As Rectangle
            Get
                Return _newRowButtonRect
            End Get

            Set(ByVal value As Rectangle)
                If NewRowButtonRect = value Then Return
                _newRowButtonRect = value
                InvalidatePanel()
            End Set
        End Property

        Private Property NewColumnButtonState As ObjectState
            Get
                Return _newColumnButtonState
            End Get

            Set(ByVal value As ObjectState)
                If NewColumnButtonState = value Then Return
                _newColumnButtonState = value
                InvalidatePanel()
            End Set
        End Property

        Private Property NewRowButtonState As ObjectState
            Get
                Return _newRowButtonState
            End Get

            Set(ByVal value As ObjectState)
                If NewRowButtonState = value Then Return
                _newRowButtonState = value
                InvalidatePanel()
            End Set
        End Property

#End Region
#Region "Paint"
        Private Sub OnTablePanelPaint(ByVal sender As Object, ByVal e As PaintEventArgs)
            CalculateRectangles()
            Using cache As GraphicsCache = New GraphicsCache(e)
                If ShowGrid Then DrawButtons(cache)
            End Using
        End Sub

        Private Sub DrawButtons(ByVal cache As GraphicsCache)
            cache.DrawImageUnscaled(CreateColumnImage, NewColumnButtonRect)
            cache.DrawImageUnscaled(CreateRowImage, NewRowButtonRect)
        End Sub

#End Region
#Region "Events"
        Private Sub OnTablePanelMouseUp(ByVal sender As Object, ByVal e As MouseEventArgs)
            CalculateObjectStates(e)
        End Sub

        Private Sub OnTablePanelMouseDown(ByVal sender As Object, ByVal e As MouseEventArgs)
            CalculateObjectStates(e)
        End Sub

        Private Sub OnTablePanelMouseMove(ByVal sender As Object, ByVal e As MouseEventArgs)
            CalculateRectangles()
            CalculateObjectStates(e)
        End Sub

        Private Sub OnTablePanelMouseClick(ByVal sender As Object, ByVal e As MouseEventArgs)
            If e.Button <> MouseButtons.Left Then Return
            If NewColumnButtonRect.Contains(e.Location) Then
                CreateColumn()
            ElseIf NewRowButtonRect.Contains(e.Location) Then
                CreateRow()
            End If
        End Sub

#End Region
#Region "Calculations"
        Private Shared ReadOnly ButtonSize As Size = ScaleUtils.ScaleValue(New Size(16, 16))

        Private Sub CalculateRectangles()
            NewColumnButtonRect = LayoutPanel.Bounds.RightTop().CreateRect(ButtonSize).WithOffset(-ButtonSize.Width, ButtonSize.Height)
            NewRowButtonRect = LayoutPanel.Bounds.RightBottom().CreateRect(ButtonSize).WithOffset(-ButtonSize.Width, -ButtonSize.Height)
        End Sub

        Private Sub CalculateObjectStates(ByVal e As MouseEventArgs)
            NewColumnButtonState = CalculateObjectStateCore(NewColumnButtonRect, e)
            NewRowButtonState = CalculateObjectStateCore(NewRowButtonRect, e)
        End Sub

        Private Function CalculateObjectStateCore(ByVal bounds As Rectangle, ByVal e As MouseEventArgs) As ObjectState
            Return If(bounds.Contains(e.Location) AndAlso (e.Button And MouseButtons.Left) = 0, ObjectState.Hot, ObjectState.Normal)
        End Function

#End Region
        Private Sub CreateColumn()
            tablePanel.Columns.Add(New TablePanelColumn(TablePanelEntityStyle.Relative, 50F))
        End Sub

        Private Sub CreateRow()
            tablePanel.Rows.Add(New TablePanelRow(TablePanelEntityStyle.Absolute, 38F))
        End Sub

        Private Sub InvalidatePanel()
            LayoutPanel.Invalidate()
        End Sub

        Private ReadOnly Property CreateColumnImage As Image
            Get
                Return imageList.GetImage(If(NewColumnButtonState = ObjectState.Hot, 1, 0), New Size(16, 16))
            End Get
        End Property

        Private ReadOnly Property CreateRowImage As Image
            Get
                Return imageList.GetImage(If(NewRowButtonState = ObjectState.Hot, 3, 2), New Size(16, 16))
            End Get
        End Property

        Public Overrides ReadOnly Property LayoutPanel As XtraLayoutPanelBase
            Get
                Return tablePanel
            End Get
        End Property

        Public ReadOnly Property TableLayoutPanel As TablePanel
            Get
                Return tablePanel
            End Get
        End Property
    End Class
End Namespace
